### ISSUE TITLE
Gradients can't be computed for keras embeddings

### ISSUE THREAD

**Issue by CloseChoice** [tensorflow/tensorflow#75351]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

No

### Source

source

### TensorFlow version

2.15.1

### Custom code

Yes

### OS platform and distribution

Windows 11, Ubuntu 22.04LTS

### Python version

3.11.6

### Current behavior?

We have a problem in shap for quite some while now with embeddings (see [here](https://github.com/shap/shap/issues/3440)). Since we manipulate the graph to adjust the gradient calculation in order to produce shap values we need the layers to be backpropagatable. This does not seem the case for `tensorflow.keras.layers.Embedding` and we do not know a way around this.

(In a previous version there was the possibility to [manipulate](https://github.com/shap/shap/blob/master/shap/explainers/_deep/deep_tf.py#L412-L416) the [`_IsBackpropagatable` function](https://github.com/tensorflow/tensorflow/blob/v1.10.0/tensorflow/python/ops/gradients_impl.py#L293) but this is no longer possible)

In the example below one can see that the gradients just become `None` if the model contains an embedding layer. 
Is there a way around this, so that we can calculate gradients for embeddings again?

### Standalone code to reproduce the issue

```python
import tensorflow as tf
from tensorflow.keras import layers, models
from tensorflow.keras.datasets import imdb
from tensorflow.keras.preprocessing.sequence import pad_sequences

# Load the IMDb dataset
max_features = 10000  # Only consider the top 10,000 words
maxlen = 100  # Only consider the first 100 words of each movie review

(X_train, y_train), (X_test, y_test) = imdb.load_data(num_words=max_features)
X_train = pad_sequences(X_train, maxlen=maxlen)
X_test = pad_sequences(X_test, maxlen=maxlen)

# Build the model
model = models.Sequential()
embedding_layer = layers.Embedding(input_dim=max_features, output_dim=128, input_length=maxlen)
model.add(embedding_layer)
flat_layer = layers.Flatten()
model.add(flat_layer)
dense_layer = layers.Dense(1, activation='sigmoid')
model.add(dense_layer)

# Build the same model except for the embedding layer
new_model = models.Sequential()
new_model.add(flat_layer)
new_model.add(layers.Dense(1, activation="sigmoid"))

# Forward pass and gradient extraction
@tf.function
def get_gradients_model(inputs):
    inputs = tf.cast(inputs, tf.float32)  # Convert inputs to float32
    with tf.GradientTape() as tape:
        tape.watch(inputs)  # Watch the input tensor to compute gradients w.r.t. it
        predictions = model(inputs)
    
    gradients = tape.gradient(predictions, inputs)
    
    return predictions, gradients

@tf.function
def get_gradients_new_model(inputs):
    inputs = tf.cast(inputs, tf.float32)  # Convert inputs to float32
    with tf.GradientTape() as tape:
        tape.watch(inputs)  # Watch the input tensor to compute gradients w.r.t. it
        predictions = new_model(inputs)
    
    gradients = tape.gradient(predictions, inputs)
    
    return predictions, gradients

# Example usage
sample_input = X_train[:1]  # Select a sample from the training set
sample_label = y_train[:1]  # Corresponding label

predictions, gradients = get_gradients_model(sample_input)
predictions2, gradients2 = get_gradients_new_model(sample_input)
print("Gradients for model:", gradients)
print("Gradients for new_model:", gradients2)
```

### Relevant log output

Gradients for model: None
Gradients for new_model: tf.Tensor(
[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
  0. 0. 0. 0.]], shape=(1, 100), dtype=float32)

**Comment by sanskarmodi8** [comment]
Hi @CloseChoice ,

I looked into your issue and found that embedding layers generally don't have direct gradients with respect to their inputs (token indices), as they map discrete indices to continuous vectors. Instead, gradients typically flow through the embedding matrix, which can cause challenges in gradient-based methods like SHAP.

### Workarounds:

1. **One-Hot Encoding**:  
   You can convert the input to one-hot encoded vectors before passing them into the embedding layer. This allows TensorFlow to compute gradients for the embeddings but increases memory usage significantly, as the input becomes much larger in terms of dimensionality.

2. **Gradient on the Embedding Matrix**:  
   Rather than trying to get gradients directly for the input tokens, you could compute the gradients with respect to the embedding matrix itself. This approach allows you to analyze how each input token influences the prediction through the learned embedding representation.

3. **SHAP KernelExplainer**:  
   If direct gradient calculation remains problematic, using SHAP’s `KernelExplainer` could be a good alternative. This method approximates SHAP values based on model predictions and doesn’t rely on gradient calculations, making it more robust for cases involving embedding layers.

**Comment by sanskarmodi8** [comment]
Hi @CloseChoice,

Thanks for the clarification. I’m sorry, am not familiar with an exact method to manipulate TensorFlow under-the-hood for calculating custom gradients specifically for embedding layers.

However, a possible approach could be to dive into TensorFlow’s lower-level APIs or custom gradient functions to explore if there's a way to achieve this.

**Comment by tilakrayal** [comment]
@CloseChoice,
I tried to execute the mentioned code on both tensorflow v2.17 & tf-nightly and observed that the output is providing zeroes on v2.17 but whereas on tf-nightly the output is provided as numericals. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/6349eb52e0b8c843f683e4e6a80d2cba/untitled2100.ipynb). Thank you!

**Comment by CloseChoice** [comment]
> @CloseChoice, I tried to execute the mentioned code on both tensorflow v2.17 & tf-nightly and observed that the output is providing zeroes on v2.17 but whereas on tf-nightly the output is provided as numericals. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/6349eb52e0b8c843f683e4e6a80d2cba/untitled2100.ipynb). Thank you!

Thanks for looking into this. The gradients still are `None` and this is my main concern. I do understand that embedding layers are not meant to calculate gradients on, but would appreciate a workaround (if one is known)

**Comment by Venkat6871** [comment]
Hi @CloseChoice ,
Thank you so much for filing this issue. We truly appreciate your support. At the moment, this request is not prioritized and mark it for community contributions. If you would like to contribute on a fix, please feel free to fork the repository and submit a pull request in accordance with the TensorFlow [CONTRIBUTION GUIDELINES](https://www.tensorflow.org/community/contribute/code) and link this issue in your PR to ensure the maintainers can easily review and collaborate.
To avoid duplicate work, please leave a comment here if you decide to take it on!

Thank you again for your interest and we look forward to seeing your contributions!

**Comment by Venkat6871** [comment]
Hi @CloseChoice ,
We are closing this issue due to inactivity. If this is still relevant and you would like to work on it, please feel free to reopen the issue or submit a pull request referencing it.

Thank you for your understanding.
